Rollback modbus to version 0.107.7 keep new functionality

janiversen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR does not solve the problems, but enables users to use modbus again, by reverting the change to async. It is basically an intelligent revert to version 0.107.7, meaning the new functionality is kept but we do not use pymodbus async.

Some of the testers have thankfully pinpointed (and reported) errors in the pymodbus asyncio implementation. It is not known when the reported bugs will be fixed, traditionally asyncio in pymodbus is not moving fast. In order to make the users happy, and revert is needed (this PR)

However the async version are to be updated to use the serial/tcp already
available instead of the flaky pymodbus version. pymodbus is still
needed to encode/decode the messages. In order to be able to share the new (not production) version with e.g. alfa testers a subdirectory have been added.

Having a subdirectory with new not production code, is not the standard way, but considering that this might take quite a time to solve, it seems to be the only way of not loosing the code and be able to work on it.

Of course test cases are also updated.

In order not to change documentation and/or configs already in place, the parameter “delay” is accepted with a warning that is currently is unused.

Before merging this change, it is important to get feedback from a couple of beta testers.

Once merged, please mark this PR to be included in the next patch release.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
fixes #34269, fixes #34046, fixes #34233, fixes #33947 
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x ] The code change is tested and works locally.
- [ x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ x] There is no commented out code in this PR.
- [ x] I have followed the [development checklist][dev-checklist]
- [x ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

janiversen committed:
Rollback modbus to version 0.107.7

Update manifest to not use async.

Rollback entities to sync version.

Keep newer modifications apart from async.

Rollback __init__ to sync version but keep the new functionality.

add async sub directory

Adding the current (not working) version in a sub directory,
to allow easy sharing with a few alfa testers.

The async version are to be updated to use the serial/tcp already
available instead of the flaky pymodbus version. pymodbus is still
needed to encode/decode the messagess.

Update test cases to reflect sync implementation, but
keep the new functionality like e.g. conftest.py.

janiversen commented:
@vzahradnik could you take a look ? I am aware that many details could have been done better, but I have tried to strike a balance between reverting 100% to 0.107.7 and keep the new functionality.

I like the bot, it added “new integration”, I did not change THAT much :-)

vzahradnik commented:
@janiversen it looks good. As for my open PRs, I think it would be best not to merge them at this point. I will keep working with the async code, and maybe finally writing some unit tests.

Please let me know if there's any progress with the `pymodbus`. Because I'm not yet in the manifest, I'm not notified about all the changes.

Thanks!

janiversen commented:
@vzahradnik thanks for looking. I agree it is better to postpone merging of your PRs. I am not sure what is best to close them or put a big note in them.

Once this PR is merged I will get started on getting the async version back to life, but with the use of non-io parts (basically framer) from pymodbus and io parts from hass.

vzahradnik commented:
@janiversen I think it would be best to put a note in them and to keep them open. They are open for a while, and it does no harm in keeping it so.

I will update them and put a reference to this PR. If maintainers decide to close them, I will create new ones when the code is ready again.

janiversen commented:
Sounds like a good plan. Apart from the “async” and “await” both interfaces (test and runtime) are the same..the changes are internal.

MartinHjelmare commented:
You can publish a branch on your github fork with the async changes if you need beta testers. I don't think we should add it here if that package is not used.

janiversen committed:
do not publish async version

The async version will be made available in a forked repo, until
it is ready to replace the production code.

janiversen commented:
Removed async in for test and ha.

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Should we convert to draft until user tests are done, to protect from merging?

janiversen commented:
That is a ver good idea, thanks

vzahradnik commented:
> Should we convert to draft until user tests are done, to protect from merging?

Great idea! Probably this would be useful for my pull requests too. They are all based on *async* code.

For a reference: PR #33642, #32439, and #33551.

MartinHjelmare commented:
Converted all three.

janiversen commented:
@MartinHjelmare thanks. Any idea when the next patch will be closed? (to get a timeframe for testing)

MartinHjelmare commented:
Not sure. If there's enough PRs on a milestone we try to release asap. Last days before beta cut we often don't do patch releases since the beta is coming up anyway.

dolenec commented:
How to test it? I have same problem but do not know how to test it.. when can I get changed file?

janiversen commented:
I am not sure what you mean, the files are in this PR. You test it by modifying your system with these files.

lopez3 commented:
Tested in my system, works fine. Fixed #34269.

dolenec commented:
Currently all seems fine.. But normaly it will take from 1 to 24 hours to stop working.. But currently is working with changed files..

Deez73 commented:
I can also confirm that this PR seems to fix the previous problems with modbus serial failing when upgraded from 0.107.7 to any 0.108.x version.

thehaxxa commented:
+1 Looks good from my end too.

janiversen commented:
@MartinHjelmare We have enough positive test results. Can you please remove the “draft” for this PR so it can be merged, thanks in advance.

dolenec commented:
I can confirm that changes made in PR is solution and is working.

iltomma commented:
Hi, I upgrade HA to 0.108.6 from 0.108.5 and serial Modbus doesn't work.
Last working HA version is 0.107.7
This is my log:
```
2020-04-18 10:11:22 ERROR (MainThread) [homeassistant.setup] Error during setup of component modbus
Traceback (most recent call last):

  File "/usr/src/homeassistant/homeassistant/setup.py", line 171, in _async_setup_component
    hass, processed_config

  File "/usr/src/homeassistant/homeassistant/components/modbus/__init__.py", line 139, in async_setup
    await hass.async_add_executor_job(start_modbus)

  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)

  File "/usr/src/homeassistant/homeassistant/components/modbus/__init__.py", line 111, in start_modbus
    client.setup()

  File "/usr/src/homeassistant/homeassistant/components/modbus/__init__.py", line 207, in setup
    loop=self._loop,

  File "/usr/local/lib/python3.7/site-packages/pymodbus/client/asynchronous/serial.py", line 75, in __new__
    yieldable = factory_class(framer=framer, port=port, **kwargs)

  File "/usr/local/lib/python3.7/site-packages/pymodbus/client/asynchronous/factory/serial.py", line 104, in async_io_factory
    client = AsyncioModbusSerialClient(port, proto_cls, framer, loop, **kwargs)

  File "/usr/local/lib/python3.7/site-packages/pymodbus/client/asynchronous/asyncio/__init__.py", line 689, in __init__
    self._connected_event = asyncio.Event()

  File "/usr/local/lib/python3.7/asyncio/locks.py", line 249, in __init__
    self._loop = events.get_event_loop()

  File "/usr/local/lib/python3.7/asyncio/events.py", line 644, in get_event_loop% threading.current_thread().name)

RuntimeError: There is no current event loop in thread 'SyncWorker_14'.
```
``

Deez73 commented:
According to change notes, this fix/PR was not part of the 0.108.6 release.
Use this PR’s files until it’s part of a .7 release or stay at 0.107.7
